src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,385 @@
+  if (this->training)
+  {
+    // Check only during training, batch-size can be one during inference.
+    if ((batchSize * higherDimension) == 1 && inputSize == 1)
+    {
+      Log::Warn << "Variance for single element isn't defined and" <<
+          " will be set to 0.0 for training. Use a batch-size" <<
+          " greater than 1 to fix the warning." << std::endl;
+    }
+
+    // Input corresponds to output from previous layer.
+    // Used a cube for simplicity.
+    arma::Cube<typename MatType::elem_type> inputTemp(
+        const_cast<MatType&>(input).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+
+    // Initialize output to same size and values for convenience.
+    arma::Cube<typename MatType::elem_type> outputTemp(
+        const_cast<MatType&>(output).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+    outputTemp = inputTemp;
+
+    // Calculate mean and variance over all channels.
+    MatType mean = arma::mean(arma::mean(inputTemp, 2), 0);
+    variance = arma::mean(arma::mean(arma::pow(
+        inputTemp.each_slice() - arma::repmat(mean,
+        inputSize, 1), 2), 2), 0);
+
+    outputTemp.each_slice() -= arma::repmat(mean, inputSize, 1);
+
+    // Used in backward propagation.
+    inputMean.set_size(arma::size(inputTemp));
+    inputMean = outputTemp;
+
+    // Normalize output.
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(variance,
+        inputSize, 1) + eps);
+
+    // Re-used in backward propagation.
+    normalized.set_size(arma::size(inputTemp));
+    normalized = outputTemp;
+
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
+
+    count += 1;
+    // Value for average factor which used to update running parameters.
+    double averageFactor = average ? 1.0 / count : momentum;
+
+    double nElements = 0.0;
+    if (input.n_elem - size != 0)
+      nElements = 1.0 / (input.n_elem - size + eps);
+
+    // Update running mean and running variance.
+    runningMean = (1 - averageFactor) * runningMean + averageFactor *
+        mean.t();
+    runningVariance = (1 - averageFactor) * runningVariance +
+       input.n_elem * nElements *
+       averageFactor * variance.t();
+  }
+  else
+  {
+    // Normalize the input and scale and shift the output.
+    output = input;
+    arma::Cube<typename MatType::elem_type> outputTemp(
+        const_cast<MatType&>(output).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+
+    outputTemp.each_slice() -= arma::repmat(runningMean.t(),
+        inputSize, 1);
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(runningVariance.t(),
+        inputSize, 1) + eps);
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
```

Comment:
Ahh, it looks like we cannot do `each_col()` here; `Cube` does not support it.  I think it should be `each_col()` since we want to apply the same normalization to each row element.  But in any case it doesn't matter because we can't do it :)